### PR TITLE
feat(kuma-cp): use ResourceIdentifier in MeshContext structs

### PR DESCRIPTION
**Pull request by lobkovilya**
* use ResourceIdentifier in mesh context and proxy objects
* change in the public API for MeshMultiZoneService status. Storing only `name` of the MeshService is not enough to uniquely identify the resource
* resolve BackendRefs in MeshTCPRoute and MeshHTTPRoute

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/resources/model/resource.go [hunk 1]
```diff
@@ -704,6 +704,26 @@ func TargetRefToResourceIdentifier(meta ResourceMeta, tr common_api.TargetRef) R
...
+	case br.Kind == common_api.MeshExternalService:
+	case br.Kind == common_api.MeshMultiZoneService:
+	default:
+		return resolved
+	}
+
+	resolved.Resource = &TypedResourceIdentifier{
+		ResourceIdentifier: TargetRefToResourceIdentifier(meta, br.TargetRef),
+		ResourceType:       ResourceType(br.Kind),
+		SectionName:        fmt.Sprintf("%d", br.Port),
```
Comment (lukidzi) [inline pkg/core/resources/model/resource.go:721]: We should check if the port is `nil`. I think it's not required for `MeshExternalService`, which might cause a panic.

File: pkg/xds/context/context.go [hunk 2]
```diff
@@ -65,8 +67,7 @@ type LabelValue struct {
 	Value string
 }
 type (
-	ServiceName                  string
-	LabelsToValuesToServiceNames map[LabelValue]map[ServiceName]bool
+	LabelsToValuesToServiceNames map[LabelValue]map[core_model.ResourceIdentifier]bool
```
Comment (lukidzi) [inline pkg/xds/context/context.go:70]: Should we rename? `LabelValuesToResourceIdentifiers`

File: pkg/xds/context/context.go [hunk 3]
```diff
@@ -78,11 +79,11 @@ type MeshContext struct {
 	Resource                                *core_mesh.MeshResource
 	Resources                               Resources
 	DataplanesByName                        map[string]*core_mesh.DataplaneResource
-	MeshServiceByName                       map[string]*meshservice_api.MeshServiceResource
+	MeshServiceByName                       map[core_model.ResourceIdentifier]*meshservice_api.MeshServiceResource
```
Comment (lukidzi) [inline pkg/xds/context/context.go:82]: Maybe this one too, `MeshServiceByIdentifiers`

File: pkg/xds/sync/dataplane_proxy_builder.go [hunk 4]
```diff
@@ -111,7 +114,25 @@ func (p *DataplaneProxyBuilder) resolveVIPOutbounds(meshContext xds_context.Mesh
 	if dataplane.Spec.Networking.GetTransparentProxying() == nil {
 		newOutbounds := []*core_xds.Outbound{}
 		for _, o := range dataplane.Spec.Networking.Outbound {
-			newOutbounds = append(newOutbounds, &core_xds.Outbound{LegacyOutbound: o})
+			if o.BackendRef != nil {
+				newOutbounds = append(newOutbounds, &core_xds.Outbound{
```
Comment (lukidzi) [inline pkg/xds/sync/dataplane_proxy_builder.go:118]: Maybe we can extract this into a new method?

File: pkg/test/xds/builders/context_builder.go [hunk 5]
```diff
@@ -22,8 +23,8 @@ func Context() *ContextBuilder {
 				Resource:                  samples.MeshDefault(),
 				EndpointMap:               map[core_xds.ServiceName][]core_xds.Endpoint{},
 				ServicesInformation:       map[string]*xds_context.ServiceInformation{},
-				MeshExternalServiceByName: map[string]*meshexternalservice_api.MeshExternalServiceResource{},
-				MeshServiceByName:         map[string]*meshservice_api.MeshServiceResource{},
+				MeshExternalServiceByName: map[model.ResourceIdentifier]*meshexternalservice_api.MeshExternalServiceResource{},
```
Comment (jakubdyszkiewicz) [inline pkg/test/xds/builders/context_builder.go:26]: Is it still by the name then?

File: pkg/xds/context/context.go [hunk 6]
```diff
@@ -101,70 +102,66 @@ type ServiceInformation struct {
...
+		} else {
+			key := core_model.TypedResourceIdentifier{
+				ResourceType: core_model.ResourceType(reachableBackend.Kind),
+				ResourceIdentifier: core_model.TargetRefToResourceIdentifier(dataplane.GetMeta(), common_api.TargetRef{
+					Name:      reachableBackend.Name,
+					Namespace: reachableBackend.Namespace,
+				}),
+			}
+			if key.Zone == "" {
+				fmt.Println("DEBUG", "zone is not set", dataplane.GetMeta().GetLabels(), dataplane.GetMeta().GetMesh(), dataplane.GetMeta().GetName())
```
Comment (jakubdyszkiewicz) [inline pkg/xds/context/context.go:133]: remove and one more a couple of lines below

File: pkg/xds/envoy/listeners/listener_builder.go [hunk 7]
```diff
@@ -47,6 +47,9 @@ func NewOutboundListenerBuilder(
 	port uint32,
 	protocol core_xds.SocketAddressProtocol,
 ) *ListenerBuilder {
+	if address == "" {
+		address = "127.0.0.1"
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/listener_builder.go:51]: where is this used? I wonder about IPV6

Comment (lobkovilya) [reply pkg/xds/envoy/listeners/listener_builder.go:51]: This is just the logic we have here https://github.com/kumahq/kuma/blob/master/api/mesh/v1alpha1/dataplane_helpers.go#L239. I can open an issue to figure this out

Comment (jakubdyszkiewicz) [reply pkg/xds/envoy/listeners/listener_builder.go:51]: I see, let's leave it as is if this results in the same behavior

**Review by jakubdyszkiewicz — APPROVED** [review]
I'm having a bit of a hard time to wrap my head around those changes. Aside of some nits it looks ok

File: pkg/xds/sync/dataplane_proxy_builder.go (hunk 4 above)
Comment (lobkovilya) [reply pkg/xds/sync/dataplane_proxy_builder.go]: I created `AsOutbounds()` method for MeshService, MeshExternalService, MeshMultiZoneService, Dataplane and it simplified the code for mesh and dpp context builders.

File: pkg/core/resources/model/resource.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/core/resources/model/resource.go]: Good catch, fixed

**Review by lukidzi — APPROVED** [review]

**Pull request merged**